src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.cs

```diff
@@ -138,37 +149,53 @@ public static void AreAlmostEqual(FileInfo expected, Rectangle expectedRect, Fil
 
 						if (cumulativeError > permittedPixelError)
 						{
-							Assert.Fail($"Difference between expected={expectedPixel} and actual={actualPixel} at {x},{y} exceeds permitted error of {permittedPixelError}");
+							Assert.Fail(WithContext(builder: builder => builder
+								.AppendLine($"Difference between pixels at ({x},{y}) exceeds permitted error of {permittedPixelError}")
+								.AppendLine($"expectedPixel: {ToArgbCode(expectedPixel)} {expectedPixel}")
+								.AppendLine($"actualPixel  : {ToArgbCode(actualPixel)} {actualPixel}")
+								.AppendLine($"permittedPixelError: {permittedPixelError}")
+								.AppendLine($"cumulativeError    : {cumulativeError}")
+							));
 						}
 					}
 			}
-		}
 
-		public static void AreNotEqual(FileInfo expected, FileInfo actual, IAppRect rect)
-			=> AreNotEqual(expected, actual, new Rectangle((int)rect.X, (int)rect.Y, (int)rect.Width, (int)rect.Height));
-		public static void AreNotEqual(FileInfo expected, FileInfo actual, Rectangle? rect = null)
-		{
-			rect = rect ?? new Rectangle(0, 0, int.MaxValue, int.MaxValue);
-			AreNotEqual(expected, rect.Value, actual, rect.Value);
+			string WithContext(string message = null, Action<StringBuilder> builder = null)
+			{
+				return new StringBuilder()
+					.AppendLine($"ImageAssert.AreAlmostEqual @ line {line}")
+					.AppendLine("permittedPixelError: " + permittedPixelError)
+					.AppendLine("expected: " + expected?.Name + (expectedRect == FirstQuadrant ? null : $"in {expectedRect}"))
+					.AppendLine("actual  : " + actual?.Name + (actualRect == FirstQuadrant ? null : $"in {actualRect}"))
+					.AppendLine("====================")
+					.ApplyIf(message != null, x => x.AppendLine(message))
+					.ApplyIf(builder != null, builder)
+					.ToString();
+			}
 		}
+		#endregion
 
-		public static void AreNotEqual(FileInfo expected, IAppRect expectedRect, FileInfo actual, IAppRect actualRect)
-			=> AreNotEqual(
-				expected,
-				new Rectangle((int)expectedRect.X, (int)expectedRect.Y, (int)expectedRect.Width, (int)expectedRect.Height),
-				actual,
-				new Rectangle((int)actualRect.X, (int)actualRect.Y, (int)actualRect.Width, (int)actualRect.Height));
+		#region AreNotEqual
+		public static void AreNotEqual(FileInfo expected, FileInfo actual, IAppRect rect, [CallerLineNumber] int line = 0)
+			=> AreNotEqualImpl(expected, rect.ToRectangle(), actual, rect.ToRectangle(), line);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to add validation of parameter 'rect' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5170992)